src/test/scala/sigmastate/lang/SigmaSpecializerTest.scala

```diff
@@ -77,4 +93,49 @@ class SigmaSpecializerTest extends PropSpec with PropertyChecks with Matchers wi
         Where(Outputs, 21, GE(ExtractAmount(TaggedBox(21)), LongConstant(10)))
   }
 
+  property("AND flattening predefined") {
+    spec("true && true") shouldBe AND(TrueLeaf, TrueLeaf)
+    spec("true && false") shouldBe AND(TrueLeaf, FalseLeaf)
+    spec("true && (true && 10 == 10)") shouldBe
+      AND(TrueLeaf, TrueLeaf, EQ(IntConstant(10), IntConstant(10)))
+    spec("true && true && true") shouldBe AND(TrueLeaf, TrueLeaf, TrueLeaf)
+    spec("true && (true && (true && true)) && true") shouldBe
+      AND(TrueLeaf, TrueLeaf, TrueLeaf, TrueLeaf, TrueLeaf)
+  }
+
+  property("AND flattening, SigmaBooleans untouched") {
+    val sigmaBooleans1 = AND(Seq(TrueLeaf, CAND(Seq(proveDlogGen.sample.get, proveDHTGen.sample.get))))
+    spec(Map(), sigmaBooleans1) shouldBe sigmaBooleans1
+    val sigmaBooleans2 = AND(Seq(TrueLeaf, COR(Seq(proveDlogGen.sample.get, proveDHTGen.sample.get))))
+    spec(Map(), sigmaBooleans2) shouldBe sigmaBooleans2
+  }
```

Comment:
@greenhat How ProveDlog and DiffieHellmanTuple sigma booleans interact with booleans during flattening? Do they flatten together?

Reply:
Sigma booleans are ignored during flattening. No. 

I just checked that we don't touch them during flattening.

Reply:
I mean if we have `AND(true, true, AND(ProveDlog(...), ProveDlog(...)))` what will happen?
I think we should flatten.
CAND, COR should be preserved however.
In current version of IR, SigmaBoolean extends Value[SBoolean.type], so they can be mixed among arguments of AND and OR

Reply:
Oh, I thought they belong to `CAND/COR` ops. My bad. I'm sure the will be flattened. Let me add the case.